docs/architecture/adr-081-protobuf-mgmt.md

```diff
@@ -0,0 +1,202 @@
+understood, are as follows.
+
+1. Tendermint needs Go code generated from `.proto` files.
+2. Consumers of Tendermint's `.proto` files want to be able to access these
+   files in a reliable and efficient way.
+3. The Tendermint Core team wants to provide stable interfaces that are as easy
+   as possible to maintain, on which consumers can depend, and to be able to
+   notify those consumers promptly when those interfaces change. To this end, we
+   want to:
+   1. Prevent any breaking changes from being introduced in minor/patch releases
+      of Tendermint. Only major version updates should be able to contain
+      breaking interface changes.
+   2. Ensure that `.proto` files don't contain invalid syntax.
+   3. Ensure that `.proto` files look reasonably similar in format for code
+      readability purposes.
+
+There was also discussion surrounding the notion of automated documentation
+generation and hosting, but it is not clear at this time whether this would be
+that valuable to any of our stakeholders. What will, of course, be valuable at
+minimum would be better documentation (in comments) of the `.proto` files
+themselves.
+
+## Alternative Approaches
+
+### Meeting stakeholders' needs
+
+1. Go stub generation from protos. We could use:
+   1. [Buf]. This approach has been rather cumbersome up to this point, and it
+      is not clear what Buf really provides beyond that which `protoc` provides
+      to justify the additional complexity in configuring Buf for stub
+      generation.
+   2. [protoc] - the Protocol Buffers compiler.
+2. Notification of breaking changes:
+   1. Buf in CI for releases only.
+   2. Buf in CI on every pull request (this was the case at the time of this
+      decision, and the team decided that the signal-to-noise ratio for this
+      approach was too low to be of value).
+3. `.proto` linting:
+   1. Buf in CI on every pull request
+4. `.proto` formatting:
+   1. [clang-format] locally and a [clang-format GitHub Action] in CI to check
+      that files are formatted properly on every pull request.
+5. Sharing of `.proto` files in a versioned, reliable manner:
+   1. Consumers could simply clone the Tendermint repository, check out a
+      specific commit, tag or branch and manually copy out all of the `.proto`
+      files they need. This requires no effort from the Tendermint Core team and
+      will continue to be an option for consumers. The drawback of this approach
+      is that it requires manual coding/scripting to implement and is brittle in
+      the face of bigger changes.
+   2. Uploading our `.proto` files to Buf's registry on every release. This is
+      by far the most seamless for consumers of our `.proto` files, but requires
+      the dependency on Buf. This has the additional benefit that the Buf
+      registry will automatically [generate and host
+      documentation][buf-docs-gen] for these protos.
+   3. We could create a process that, upon release, creates a `.zip` file
```

Comment:
This is a very simple workflow. GitHub already provides a pre-baked upload-artifact workflow, so all we'd have to script is creating the file, e.g., `zip -9vr protobuf.zip tendermint/proto` or words to that effect.

Reply:
If someone is fetching a packaged artifact from a tagged release, they will automatically get the right version of the proto files. I'm suggesting we'd set this up to be published as a release artifact. We definitely shouldn't leave people to guess.

Reply:
Of course, people would definitely not be guessing as to which version they're retrieving and the github URL would (I think) contain the tag / release version so it would be written down clearly in consumer's repos if they elect to script the fetch. I'm mostly just trying to elaborate that users would each have to role their own scripting and part of that scripting would be making sure that they did the version pin correctly. Perhaps not a massive complication, but one that buf at least notionally handles for you.